src/asm.rs

```diff
@@ -1,254 +1,544 @@
-                                continue;
-                            },
-                        };
-                    operand_numbers.insert(idx, current_number);
-                    current_number += 1;
-                    let var = self.current_func().new_local(None, ty, "output_register");
-                    output_vars.insert(idx, var);
+
+                InlineAsmOperandRef::In { reg, value } => {
+                    if let Ok(constraint) = reg_to_gcc(reg).map(Cow::Borrowed) {
+                        inputs.push(AsmInOperand { 
+                            constraint, 
+                            rust_idx, 
+                            val: value.immediate()
+                        });
+                    } else {
+                        // left for the next pass
+                        continue
+                    }
+                }
+
+                InlineAsmOperandRef::InOut { reg, late, in_value, out_place } => {
+                    let constraint = if let Ok(constraint) = reg_to_gcc(reg) {
+                        constraint
+                    } else {
+                        // left for the next pass
+                        continue
+                    };
+
+                    // Rustc frontend guarantees that input and output types are "compatible",
+                    // so we can just use input var's type for the output variable.
+                    //
+                    // This decision is also backed by the fact that LLVM needs in and out 
+                    // values to be of *exactly the same type*, not just "compatible". 
+                    // I'm not sure if GCC is so picky too, but better safe than sorry.
+                    let ty = in_value.layout.gcc_type(self.cx, false);
+                    let tmp_var = self.current_func().new_local(None, ty, "output_register");
+
+                    // If the out_place is None (i.e `inout(reg) var` syntax was used), we translate
+                    // it to one "readwrite (+) output variable", otherwise we translate it to two 
+                    // "out and tied in" vars as described above.
+                    if out_place.is_none() {
+                        outputs.push(AsmOutOperand {
+                            constraint, 
+                            rust_idx,
+                            late,
+                            readwrite: true,
+                            tmp_var, 
+                            out_place,
+                        });
+                    } else {
+                        outputs.push(AsmOutOperand {
+                            constraint, 
+                            rust_idx,
+                            late,
+                            readwrite: false,
+                            tmp_var, 
+                            out_place,
+                        });
+
+                        let out_gcc_idx = outputs.len() - 1;
+                        let constraint = Cow::Owned(out_gcc_idx.to_string());
```

Comment:
It seems like the constraint field is only ever assigned a scalar value (either a single char or an integer). Do you think it would be worth using another type than `Cow<str>` to avoid the allocation in this case.

Reply:
The LLVM probably has the same issue, so that was probably already looked at!